packages/core-contracts/src/contracts/arks/Psm3ERC4626Ark.sol

```diff
@@ -0,0 +1,249 @@
+        address expectedAsset = shouldStake ? address(susds) : address(usds);
+        if (address(erc4626Vault.asset()) != expectedAsset) {
+            revert ERC4626AssetMismatch();
+        }
+    }
+
+    /**
+     * @notice Internal function to get the total assets that are withdrawable
+     * @return withdrawableAssets Amount of assets that can be withdrawn
+     */
+    function _withdrawableTotalAssets()
+        internal
+        view
+        override
+        returns (uint256 withdrawableAssets)
+    {
+        uint256 _totalAssets = totalAssets();
+        if (_totalAssets > 0) {
+            uint256 psmUsdcBalance = config.asset.balanceOf(psm.pocket());
+            withdrawableAssets = _totalAssets < psmUsdcBalance
+                ? _totalAssets
+                : psmUsdcBalance;
+        }
+    }
+
+    /**
+     * @notice Handles the boarding process for USDS
+     * @param usdsAmount Amount of USDS to deposit
+     */
+    function _handleUsdsBoarding(uint256 usdsAmount) internal {
+        usds.forceApprove(address(erc4626Vault), usdsAmount);
+        erc4626Vault.deposit(usdsAmount, address(this));
+    }
+
+    /**
+     * @notice Handles the boarding process for sUSDS
+     * @param susdsAmount Amount of sUSDS to deposit
+     */
+    function _handleSusdsBoarding(uint256 susdsAmount) internal {
+        susds.forceApprove(address(erc4626Vault), susdsAmount);
+        erc4626Vault.deposit(susdsAmount, address(this));
+    }
+
+    /**
+     * @notice Handles the disembarking process for USDS
+     * @param usdsAmount Amount of USDS to withdraw
+     */
+    function _handleUsdsDisembarking(uint256 usdsAmount) internal {
+        erc4626Vault.withdraw(usdsAmount, address(this), address(this));
+        usds.forceApprove(address(psm), usdsAmount);
+    }
+
+    /**
+     * @notice Handles the disembarking process for sUSDS
+     * @param susdsAmount Amount of sUSDS to withdraw
+     */
+    function _handleSusdsDisembarking(uint256 susdsAmount) internal {
+        erc4626Vault.withdraw(susdsAmount, address(this), address(this));
+        susds.forceApprove(address(psm), susdsAmount);
+    }
+
+    /**
+     * @notice Handles the boarding process
+     * @param amount Amount of USDC to board
+     */
+    function _board(uint256 amount, bytes calldata) internal override {
+        // Approve PSM to take USDC
+        config.asset.forceApprove(address(psm), amount);
+
+        // Preview swap to get expected token amount
+        uint256 expectedTokenAmount = psm.previewSwapExactIn(
+            address(config.asset),
+            address(shouldStake ? susds : usds),
+            amount
+        );
+
+        // Perform swap with exact output as preview
+        uint256 tokenAmount = psm.swapExactIn(
+            address(config.asset),
+            address(shouldStake ? susds : usds),
+            amount,
+            expectedTokenAmount,
+            address(this),
+            0
+        );
+
+        if (shouldStake) {
```

Comment:
Opinion: we could reverse this and use the `_underlying` function mentioned before to fetch the asset as an IERC20 and then pass it to `_handleBoarding` that uses the IERC20 interface to do the boarding.